### PR TITLE
Add support for inner product in ANN, custom scoring and painless

### PR DESCRIPTION
**Pull request by chenqi0805**
*Issue #, if available:*
#114 #171 

*Description of changes:*
This PR
(1) Add negdotprod similarity measure
(2) extend KNN scorer to cover negative similarity measures.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java [hunk 1]
```diff
@@ -123,5 +123,13 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
     public boolean isCacheable(LeafReaderContext context) {
         return true;
     }
+
+    public static float scoreFunc(float score) {
```
Comment (jmazanec15) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java]: nit: Could you just name this score?

**Comment by jmazanec15** [comment]
@chenqi0805 Could you change user exposed names to InnerProduct? I think that would make more sense. For calls to nmslib, we could translate inner_product -> negdotprod. What do you think?

**Comment by chenqi0805** [comment]
@jmazanec15 Sure. ~I will check the places where it is user exposed~. Will open a separate issue to track the translation.

**Review by jmazanec15 — APPROVED** [review]
This looks good to me @chenqi0805! Thanks for making this change!

File: src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java [hunk 2]
```diff
@@ -123,5 +123,13 @@ public Scorer scorer(LeafReaderContext context) throws IOException {
     public boolean isCacheable(LeafReaderContext context) {
         return true;
     }
+
+    public static float score(float score) {
+        if (score >= 0) {
+            return 1 / (1 + score);
+        } else {
+            return 2 + 1 / (score - 1);
```
Comment (VijayanB) [inline src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java]:
```
private static float normalizeScore(float score) {
    if (score >= 0)
         return 1 / ( 1 + score );
   return 2 + ( 1 / (score - 1) );
   ```

Comment (VijayanB) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java]: Thanks for making changes. 
1. Do we need this to be public? 
2. Shall we add parenthesis around formula? For better readability, i am thinking we don't want reader to remember the precedence logic to understand the formula. 
What do you think? @chenqi0805

Comment (chenqi0805) [reply src/main/java/com/amazon/opendistroforelasticsearch/knn/index/KNNWeight.java]: > 1. Do we need this to be public?

Sorry, I missed the access modifier in your previous comment. The reason I maintained it public is to reuse it in `KNNScoringSpace`. If made private, we will need to duplicate the logic there. Let me know what you think.

> 2. Shall we add parenthesis around formula? For better readability, i am thinking we don't want reader to remember the precedence logic to understand the formula.
What do you think? @chenqi0805

Sure. Makes sense. Will add the parenthesis.

**Review by VijayanB — COMMENTED** [review]
LGTM! Thanks for making changes!

**Review by VijayanB — APPROVED** [review]

**Pull request merged**